bookkeeper-server/src/main/java/org/apache/bookkeeper/client/api/LedgerMetadata.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.bookkeeper.client.api;
+
+import java.io.Serializable;
+import java.util.List;
+import java.util.Map;
+import java.util.NavigableMap;
+import org.apache.bookkeeper.common.annotation.InterfaceAudience.LimitedPrivate;
+import org.apache.bookkeeper.common.annotation.InterfaceStability.Unstable;
+import org.apache.bookkeeper.net.BookieSocketAddress;
+
+/**
+ * Represents the client-side metadata of a ledger. It is immutable.
+ *
+ * @since 4.6
+ */
+@LimitedPrivate
+@Unstable
+public interface LedgerMetadata extends Serializable {
+
+    /**
+     * Returns the ensemble size of this ledger.
+     *
+     * @return the ensemble size of this ledger.
+     */
+    int getEnsembleSize();
+
+    /**
+     * Returns the write quorum size of this ledger.
+     *
+     * @return the write quorum size of this ledger.
+     */
+    int getWriteQuorumSize();
+
+    /**
+     * Returns the ack quorum size of this ledger.
+     *
+     * @return the ack quorum size of this ledger.
+     */
+    int getAckQuorumSize();
+
+    /**
+     * Returns the last entry id of this ledger.
+     *
+     * <p>If this ledger is not sealed {@link #isSealed()}, it returns {@code -1L}.
+     *
+     * @return the last entry id of this ledger if it is sealed, otherwise -1.
+     */
+    long getLastEntryId();
+
+    /**
+     * Returns the length of this ledger.
+     *
+     * <p>If this ledger is not sealed {@link #isSealed()}, it returns {@code 0}.
+     *
+     * @return the length of this ledger if it is sealed, otherwise 0.
+     */
+    long getLength();
```

Comment:
Better to return Optional<Long> for this and getLastEntryId. These won't be called a lot, so garbage isn't an issue, and it'll be clearer that the value may not exist.

Reply:
I am not sure this should be Optional. there is no point for Optional. If no data is written to a ledger, the length is 0 and last entry id is -1L.

Reply:
But the length may not be 0 and last entry Id may not be -1L in the case that the ledger has been written to, just not sealed. The Optional is to indicate that the returned values are not valid. 0 and 1L are valid values, in the case of a empty ledger being sealed.

Reply:
It is not actually an optional field. I am not sure we really need. I would prefer sticking to current interface, because both methods are already there in the old implementation and used by applications.

Reply:
It is optional if you can have a LedgerMetadata object with them unset, which is the case if the ledger is unsealed. The new api gives us a chance to break old interfaces, so we should take that opportunity to make things cleaner.

Reply:
- I am not sure whether optional is more cleaner and returning 0 is more cleaner. looking into storage apis, returning an integer value is very common and well understood for length, last value.

- I don't think it is Optional. Optional means this field is not optional or can be missing. In this case, it just mean the length is not updated yet.